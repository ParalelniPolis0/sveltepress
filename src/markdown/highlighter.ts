import { readFileSync } from 'fs'
import { resolve } from 'path'
import { fileURLToPath } from 'url'
import { getHighlighter } from 'shiki'
import type { Highlighter } from '@svelte-press/vite'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

export const COMMAND_RE = /\/\/ \[svp\! (hl(:-?\d+(,-?\d+)?)?)\]/

const nightOwl = JSON.parse(readFileSync(resolve(__dirname, './night-owl.json'), 'utf-8'))
const vitesseLight = JSON.parse(readFileSync(resolve(__dirname, './vitesse-light.json'), 'utf-8'))

const createHighlightWith: (theme: string) => Highlighter = theme => (code, lang) => getHighlighter({
  theme,
  langs: ['svelte', 'sh', 'js', 'html', 'ts'],
}).then(
  shikiHighlighter => shikiHighlighter
    .codeToHtml(code, { lang })
    .replace(/\{/g, '&#123;')
    .replace(/\}/g, '&#125;'),
)

const highlighterDark = createHighlightWith(nightOwl)

const highlighterLight = createHighlightWith(vitesseLight)

export function getCommand(line: string) {
  const matches = COMMAND_RE.exec(line)
  if (matches && matches.length) {
    const [comment] = matches
    return [comment.replace(/^\/\/ \[svp\! /, '').replace(/\]$/, ''), line.replace(COMMAND_RE, '')]
  }

  return ['', line]
}

export function highlightLine(startEnd: string, idx: number) {
  if (!startEnd)
    return `<div class="absolute left-0 right-0 z-2 h-[1.5em] bg-[rgba(0,0,0,0.1)] dark:bg-[rgba(255,255,255,0.1)] svp-code-highlight-line" style="top: calc(${idx * 1.5}em + 12px);"></div>`
}

const highlighter: Highlighter = async (code, lang) => {
  const commandDoms = []
  if (lang !== 'md') {
    code = code.split('\n').map((line, i) => {
      const [command, newLine] = getCommand(line)
      if (!command)
        return line

      const [name, params] = command.split(':')
      switch (name) {
        case 'hl':
          commandDoms.push(highlightLine(params, i))
      }
      return newLine
    }).join('\n')
  }
  return `<div class="relative bg-white dark:bg-[#011627] p-[12px] rounded text-[14px]">
      ${commandDoms.join('\n')}
      ${await highlighterLight(code, lang)}
      ${await highlighterDark(code, lang)}
      <div class="absolute top-2 right-3 text-cool-gray-3 dark:text-cool-gray-7 text-[12px] z-2">
        ${lang}
      </div>
  </div>`
}

export default highlighter
