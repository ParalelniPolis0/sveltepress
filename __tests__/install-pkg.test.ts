import { describe, expect, it } from 'vitest'
import { compile } from 'mdsvex'
import installPkg, { pkgRe } from '../src/markdown/install-pkg'
import highlighter from '../src/markdown/highlighter'

describe('install-pkg', () => {
  it('re', () => {
    expect(pkgRe.exec('@install-pkg(foo-bar)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar)",
        "foo-bar",
        undefined,
      ]
    `)
    expect(pkgRe.exec('@install-pkg(foo-bar,create)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar,create)",
        "foo-bar,create",
        undefined,
      ]
    `)
  })

  it('md', async () => {
    const source = '@install-pkg(@sveltepress/vite,create)'

    const { code } = await compile(source, {
      filename: 'install-pkg.md',
      extensions: ['.md'],
      highlight: {
        highlighter,
      },
      remarkPlugins: [installPkg],
    }) || { code: '' }
    expect(code).toMatchSnapshot()
  })
})
