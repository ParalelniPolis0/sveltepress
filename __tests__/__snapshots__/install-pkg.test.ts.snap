// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`install-pkg > md 1`] = `
"<InstallPkg><div slot=\\"npm\\">
<div class=\\"svp-code-block-wrapper\\">
  <div class=\\"svp-code-block\\">
    
    <pre class=\\"shiki vitesse-light\\" style=\\"background-color: #ffffff\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #59873A\\">npm</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">create</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">@sveltepress/vite</span></span></code></pre>
    <pre class=\\"shiki night-owl\\" style=\\"background-color: #011627\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #82AAFF; font-style: italic\\">npm</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">create</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">@sveltepress/vite</span></span></code></pre>
    <div class=\\"svp-code-block--lang\\">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div><div slot=\\"yarn\\">
<div class=\\"svp-code-block-wrapper\\">
  <div class=\\"svp-code-block\\">
    
    <pre class=\\"shiki vitesse-light\\" style=\\"background-color: #ffffff\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #59873A\\">yarn</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">create</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">@sveltepress/vite</span></span></code></pre>
    <pre class=\\"shiki night-owl\\" style=\\"background-color: #011627\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #82AAFF; font-style: italic\\">yarn</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">create</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">@sveltepress/vite</span></span></code></pre>
    <div class=\\"svp-code-block--lang\\">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div><div slot=\\"pnpm\\">
<div class=\\"svp-code-block-wrapper\\">
  <div class=\\"svp-code-block\\">
    
    <pre class=\\"shiki vitesse-light\\" style=\\"background-color: #ffffff\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #59873A\\">pnpm</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">create</span><span style=\\"color: #393A34\\"> </span><span style=\\"color: #B56959\\">@sveltepress/vite</span></span></code></pre>
    <pre class=\\"shiki night-owl\\" style=\\"background-color: #011627\\" tabindex=\\"0\\"><code><span class=\\"line\\"><span style=\\"color: #82AAFF; font-style: italic\\">pnpm</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">create</span><span style=\\"color: #D6DEEB\\"> </span><span style=\\"color: #ECC48D\\">@sveltepress/vite</span></span></code></pre>
    <div class=\\"svp-code-block--lang\\">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div></InstallPkg>"
`;
